### PR TITLE
Go 1.16 support and some general improvements

### PR DESCRIPTION
**Pull request by daenney**
* Move go-neb to be built by 1.16 and update the Dockerfile to use newer bases
  Just for general house keeping purposes and getting access to new features in go
* Replace the deprecated golint with staticcheck and fix the linter warnings
  Golint has been [deprecated and frozen](https://github.com/golang/go/issues/38968)
* Fix the sqlite3 package version which also resolves a build issue on newer gcc's
  The v2 tags were erroneous, move us back to v1 tags
* Migrate CI from Travis to Github Actions
  We've started to use GHA for more things on matrix-org, this moves neb in the same direction. The one thing we lose now is the Travis-CI webhook notifications, but given GHA will poke the author of the PR if their builds fail that feels like an acceptable trade-off. You can see what a build ends up looking like [over here](https://github.com/daenney/go-neb/runs/2618457592?check_suite_focus=true)
* Upgrade to the current release of mautrix-go
  We need this in order to be able to build with newer versions of Go

### CODE REVIEW

**Comment by daenney** [comment]
Though all the tests pass, one test is logging a panic:

```
=== RUN   TestRespondToEcho
time="2021-05-19T08:55:31Z" level=info msg="Incoming request" req.id=smW335jjZJSf req.method=POST req.path=/admin/configureClient
time="2021-05-19T08:55:31Z" level=warning msg="Device ID is not set which will result in E2E encryption/decryption not working"
time="2021-05-19T08:55:31Z" level=error msg="Failed to load next_batch token" error="sql: no rows in result set" user_id="@link:hyrule"
time="2021-05-19T08:55:31Z" level=info msg="Created new client" auto_join_rooms=true device_id= since= sync=true user_id="@link:hyrule"
time="2021-05-19T08:55:31Z" level=info msg="Responding (372 bytes)" code=200 req.id=smW335jjZJSf req.method=POST req.path=/admin/configureClient
time="2021-05-19T08:55:31Z" level=info msg="Incoming request" req.id=Bk1hytf2pIFO req.method=POST req.path=/admin/configureService
time="2021-05-19T08:55:31Z" level=info msg="Incoming configure service request" req.id=Bk1hytf2pIFO req.method=POST req.path=/admin/configureService service_id=test_echo_service service_type=echo service_user_id="@link:hyrule"
time="2021-05-19T08:55:31Z" level=info msg="Responding (72 bytes)" code=200 req.id=Bk1hytf2pIFO req.method=POST req.path=/admin/configureService
time="2021-05-19T08:55:31Z" level=error msg="Fatal Sync() error" error="ProcessResponse panicked! since= panic=RoundTrip: Unhandled request: POST /_matrix/client/r0/keys/upload\nHandlers: 2\ngoroutine 12 [running]:\nruntime/debug.Stack(0xc000706e30, 0xba9080, 0xc0001aaa20)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/runtime/debug/stack.go:24 +0x9f\nmaunium.net/go/mautrix.(*DefaultSyncer).ProcessResponse.func1(0x0, 0x0, 0xc000707d18)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/sync.go:133 +0x76\npanic(0xba9080, 0xc0001aaa20)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/runtime/panic.go:965 +0x1b9\ngithub.com/matrix-org/go-neb.(*matrixTripper).RoundTrip(0xc0000aaa78, 0xc000112300, 0xc0000aaa78, 0x0, 0x0)\n\t/home/runner/work/go-neb/go-neb/testutil_test.go:43 +0x33f\nnet/http.send(0xc000112300, 0xd96080, 0xc0000aaa78, 0x0, 0x0, 0x0, 0xc0000102c8, 0x1e, 0x1, 0x0)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/net/http/client.go:251 +0x454\nnet/http.(*Client).send(0xc0001a9890, 0xc000112300, 0x0, 0x0, 0x0, 0xc0000102c8, 0x0, 0x1, 0x419ffb)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/net/http/client.go:175 +0xff\nnet/http.(*Client).do(0xc0001a9890, 0xc000112300, 0x0, 0x0, 0x0)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/net/http/client.go:717 +0x45f\nnet/http.(*Client).Do(...)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/net/http/client.go:585\nmaunium.net/go/mautrix.(*Client).executeCompiledRequest(0xc0004441a0, 0xc000112300, 0x0, 0xee6b2800, 0xb7c920, 0xc0000102b8, 0x0, 0x0, 0x0, 0x0, ...)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:393 +0xe7\nmaunium.net/go/mautrix.(*Client).MakeFullRequest(0xc0004441a0, 0xcb0dfd, 0x4, 0xc000784c00, 0x2f, 0x0, 0xb8d6e0, 0xc00045b6f0, 0x0, 0x0, ...)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:360 +0x1e5\nmaunium.net/go/mautrix.(*Client).MakeRequest(...)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:284\nmaunium.net/go/mautrix.(*Client).UploadKeys(0xc0004441a0, 0xc00045b6f0, 0x1a, 0xc00045b700, 0x1)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:1184 +0x258\nmaunium.net/go/mautrix/crypto.(*OlmMachine).ShareKeys(0xc00048c000, 0x0, 0x4e, 0xc00045a9d0)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/crypto/machine.go:421 +0x1a5\nmaunium.net/go/mautrix/crypto.(*OlmMachine).ProcessSyncResponse(0xc00048c000, 0xc00046c4d0, 0x0, 0x0, 0x767d18)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/crypto/machine.go:180 +0x305\ngithub.com/matrix-org/go-neb/clients.(*BotClient).syncCallback(0xc0001a5220, 0xc00046c4d0, 0x0, 0x0, 0xcf5f01)\n\t/home/runner/work/go-neb/go-neb/clients/bot_client.go:113 +0x75\nmaunium.net/go/mautrix.(*DefaultSyncer).ProcessResponse(0xc000296e60, 0xc00046c4d0, 0x0, 0x0, 0x0, 0x0)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/sync.go:138 +0x110\nmaunium.net/go/mautrix.(*Client).SyncWithContext(0xc0004441a0, 0xda3a50, 0xc0000a6010, 0x0, 0x0)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:245 +0x22d\nmaunium.net/go/mautrix.(*Client).Sync(...)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:201\ngithub.com/matrix-org/go-neb/clients.(*BotClient).Sync(0xc0001a5220)\n\t/home/runner/work/go-neb/go-neb/clients/bot_client.go:170 +0x185\ncreated by github.com/matrix-org/go-neb/clients.(*Clients).initClient\n\t/home/runner/work/go-neb/go-neb/clients/clients.go:417 +0xd65\n" user_id="@link:hyrule"
```

I don't really know why, it seems the handler that gets installed for `keys/upload` somehow goes missing during the test. Since the tests pass, I'm not actually sure what to make of it

**Comment by daenney** [comment]
@Kegsay @neilalexander Could I bother you for a review on this?

**Review by neilalexander — COMMENTED** [review]
Might be better to use `golangci-lint`, which includes `staticcheck`. There's a sample config [in the Dendrite repository](https://github.com/matrix-org/dendrite/blob/master/.golangci.yml), if we're looking to settle on a standard.

**Comment by daenney** [comment]
> Might be better to use `golangci-lint`, which includes `staticcheck`. There's a sample config [in the Dendrite repository](https://github.com/matrix-org/dendrite/blob/master/.golangci.yml), if we're looking to settle on a standard.

I was thinking about that, but due to the amount of extra linters that it has and the state of go-neb that becomes a lot of extra things to fix. I'm happy to do that, but I'd prefer to do it as a follow-up PR.

File: clients/bot_client.go [hunk 1]
```diff
@@ -314,5 +314,5 @@ func (botClient *BotClient) ForwardRoomKeyToDevice(userID id.UserID, deviceID id
 		},
 	}
 
-	return botClient.olmMachine.SendEncryptedToDevice(device, forwardedRoomKey)
+	return botClient.olmMachine.SendEncryptedToDevice(device, mevt.ToDeviceForwardedRoomKey, forwardedRoomKey)
```
Comment (daenney) [inline clients/bot_client.go:317]: I'm not sure this is the right event type actually, or if this should be `ToDeviceRoomKey`. I went with `ToDeviceForwardedRoomKey` due to the `forwardedRoomKey` variable.

Comment (tulir) [reply clients/bot_client.go:317]: It's correct, the old version only allowed sending forwarded room key events (and I forgot to mention the parameter being added in changelogs, now specified in the [v0.9.7](https://github.com/tulir/mautrix-go/releases/tag/v0.9.7) notes)

**Comment by Mic92** [comment]
I realized you also incorporate my changes: https://github.com/matrix-org/go-neb/pull/351/files

**Comment by daenney** [comment]
@neilalexander @kegsay Could I get another review and/or sign-off please? I'd like to move this forward.

File: services/github/github_webhook.go [hunk 2]
```diff
@@ -368,37 +368,6 @@ func (s *WebhookService) deleteHook(owner, repo string) error {
 	return err
 }
 
-func sameRepos(a *WebhookService, b *WebhookService) bool {
```
Comment (kegsay) [inline services/github/github_webhook.go:371]: Why was this deleted? Unusued?

**Review by kegsay — APPROVED** [review]

File: services/github/github_webhook.go (hunk 2 above)
Comment (daenney) [reply services/github/github_webhook.go:371]: Yes, there's nothing using it.

```
$ ag sameRepos
services/github/github_webhook.go
371:func sameRepos(a *WebhookService, b *WebhookService) bool {
```

The only thing that mentions it is the function definition itself. I can't find anything that calls it.

Comment (daenney) [reply services/github/github_webhook.go:371]: > func sameRepos is unused (U1000) go-staticcheck

**Comment by daenney** [comment]
> Though all the tests pass, one test is logging a panic:
> 
> ```
> === RUN   TestRespondToEcho
> time="2021-05-19T08:55:31Z" level=info msg="Incoming request" req.id=smW335jjZJSf req.method=POST req.path=/admin/configureClient
> time="2021-05-19T08:55:31Z" level=warning msg="Device ID is not set which will result in E2E encryption/decryption not working"
> time="2021-05-19T08:55:31Z" level=error msg="Failed to load next_batch token" error="sql: no rows in result set" user_id="@link:hyrule"
> time="2021-05-19T08:55:31Z" level=info msg="Created new client" auto_join_rooms=true device_id= since= sync=true user_id="@link:hyrule"
> time="2021-05-19T08:55:31Z" level=info msg="Responding (372 bytes)" code=200 req.id=smW335jjZJSf req.method=POST req.path=/admin/configureClient
> time="2021-05-19T08:55:31Z" level=info msg="Incoming request" req.id=Bk1hytf2pIFO req.method=POST req.path=/admin/configureService
> time="2021-05-19T08:55:31Z" level=info msg="Incoming configure service request" req.id=Bk1hytf2pIFO req.method=POST req.path=/admin/configureService service_id=test_echo_service service_type=echo service_user_id="@link:hyrule"
> time="2021-05-19T08:55:31Z" level=info msg="Responding (72 bytes)" code=200 req.id=Bk1hytf2pIFO req.method=POST req.path=/admin/configureService
> time="2021-05-19T08:55:31Z" level=error msg="Fatal Sync() error" error="ProcessResponse panicked! since= panic=RoundTrip: Unhandled request: POST /_matrix/client/r0/keys/upload\nHandlers: 2\ngoroutine 12 [running]:\nruntime/debug.Stack(0xc000706e30, 0xba9080, 0xc0001aaa20)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/runtime/debug/stack.go:24 +0x9f\nmaunium.net/go/mautrix.(*DefaultSyncer).ProcessResponse.func1(0x0, 0x0, 0xc000707d18)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/sync.go:133 +0x76\npanic(0xba9080, 0xc0001aaa20)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/runtime/panic.go:965 +0x1b9\ngithub.com/matrix-org/go-neb.(*matrixTripper).RoundTrip(0xc0000aaa78, 0xc000112300, 0xc0000aaa78, 0x0, 0x0)\n\t/home/runner/work/go-neb/go-neb/testutil_test.go:43 +0x33f\nnet/http.send(0xc000112300, 0xd96080, 0xc0000aaa78, 0x0, 0x0, 0x0, 0xc0000102c8, 0x1e, 0x1, 0x0)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/net/http/client.go:251 +0x454\nnet/http.(*Client).send(0xc0001a9890, 0xc000112300, 0x0, 0x0, 0x0, 0xc0000102c8, 0x0, 0x1, 0x419ffb)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/net/http/client.go:175 +0xff\nnet/http.(*Client).do(0xc0001a9890, 0xc000112300, 0x0, 0x0, 0x0)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/net/http/client.go:717 +0x45f\nnet/http.(*Client).Do(...)\n\t/opt/hostedtoolcache/go/1.16.4/x64/src/net/http/client.go:585\nmaunium.net/go/mautrix.(*Client).executeCompiledRequest(0xc0004441a0, 0xc000112300, 0x0, 0xee6b2800, 0xb7c920, 0xc0000102b8, 0x0, 0x0, 0x0, 0x0, ...)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:393 +0xe7\nmaunium.net/go/mautrix.(*Client).MakeFullRequest(0xc0004441a0, 0xcb0dfd, 0x4, 0xc000784c00, 0x2f, 0x0, 0xb8d6e0, 0xc00045b6f0, 0x0, 0x0, ...)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:360 +0x1e5\nmaunium.net/go/mautrix.(*Client).MakeRequest(...)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:284\nmaunium.net/go/mautrix.(*Client).UploadKeys(0xc0004441a0, 0xc00045b6f0, 0x1a, 0xc00045b700, 0x1)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:1184 +0x258\nmaunium.net/go/mautrix/crypto.(*OlmMachine).ShareKeys(0xc00048c000, 0x0, 0x4e, 0xc00045a9d0)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/crypto/machine.go:421 +0x1a5\nmaunium.net/go/mautrix/crypto.(*OlmMachine).ProcessSyncResponse(0xc00048c000, 0xc00046c4d0, 0x0, 0x0, 0x767d18)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/crypto/machine.go:180 +0x305\ngithub.com/matrix-org/go-neb/clients.(*BotClient).syncCallback(0xc0001a5220, 0xc00046c4d0, 0x0, 0x0, 0xcf5f01)\n\t/home/runner/work/go-neb/go-neb/clients/bot_client.go:113 +0x75\nmaunium.net/go/mautrix.(*DefaultSyncer).ProcessResponse(0xc000296e60, 0xc00046c4d0, 0x0, 0x0, 0x0, 0x0)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/sync.go:138 +0x110\nmaunium.net/go/mautrix.(*Client).SyncWithContext(0xc0004441a0, 0xda3a50, 0xc0000a6010, 0x0, 0x0)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:245 +0x22d\nmaunium.net/go/mautrix.(*Client).Sync(...)\n\t/home/runner/go/pkg/mod/maunium.net/go/mautrix@v0.9.12/client.go:201\ngithub.com/matrix-org/go-neb/clients.(*BotClient).Sync(0xc0001a5220)\n\t/home/runner/work/go-neb/go-neb/clients/bot_client.go:170 +0x185\ncreated by github.com/matrix-org/go-neb/clients.(*Clients).initClient\n\t/home/runner/work/go-neb/go-neb/clients/clients.go:417 +0xd65\n" user_id="@link:hyrule"
> ```
> 
> I don't really know why, it seems the handler that gets installed for `keys/upload` somehow goes missing during the test. Since the tests pass, I'm not actually sure what to make of it

@kegsay Do you have any thoughts on this? I'm still kinda puzzled by the test. It feels wrong that it's logging a panic there.

**Comment by kegsay** [comment]
Let's leave it for now, but if you can file an issue for it that'd be great.
